### PR TITLE
Improve parent block selector button UI

### LINKED ISSUE #23766: Improve usability of parent block selector button
## The issue

The parent block selector button is only shown on hover or when you tab backwards into it. This is obviously not ideal for usability, and often, it results in people not even realizing (or just forgetting) it exists in the first place.

image

Simultaneously, one common issue with nested blocks is making it clear that you're even in a nested block in the first place. When editing Navigation Link blocks, it would be handy if there was a quick and **obvious** way to get back to the parent Navigation block. The same issue applies to Buttons, Social Icons, and in the future it could apply to the Quote block, if that ever gets updated to use nested blocks.

There's also an issue of what happens when the toolbar goes sticky at the top of the editor. The parent selector button ends up having to overlap the top toolbar, or else it isn't accessible. Previously, the block toolbar had a lower z-index than the top toolbar, but because of the parent-selector, it had to be increased, resulting in this regression when the selected block is scrolled off-screen:

image

The block toolbar shouldn't be visible at this point, but it has to be in order for the parent-selector to be clickable when the block toolbar is sticking under the top toolbar.

## Proposed solution

I propose that the parent block selector button be made always visible when a child block is selected, and that it be shown to the left of the toolbar, rather than above. This would solve all of the aforementioned issues. Here's a rough mockup:

image

### PR DESCRIPTION
**Pull request by ZebulanStanphill**
## Description
Closes #23766.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ZebulanStanphill** [comment]
Here's what I've got so far:
image

I was planning to add a gap between the parent-selector and the rest of the toolbar:

image

But adding that extra space is a lot harder than it may seem. The outermost borders and background color are provided by the `block-editor-block-contextual-toolbar` that contains both the parent-selector button as well as all the other toolbar buttons. To style it like the mockup requires a lot of style overriding.

I could use some help here to figure out what to do. Is it okay to not have the gap?

It's worth noting that something like a thicker border would be a lot easier to do, albeit a bit non-standard. Then again, separating the parent-selector with a gap even though it's technically still part of the toolbar is also non-standard.

image

So what do you think I should do? Personally, I'm okay with the current design in this PR.

**Comment by mapk** [comment]
This is currently what I'm seeing right now:

Screen Shot 2020-07-09 at 1 55 16 PM

The parent icon does not have a right border. Looking at your mockups though, I still prefer the detached parent icon. The disconnection of the two help identify it as something tangentially related to the selected block, and not as part of the selected block itself.

**Comment by ZebulanStanphill** [comment]
Well, I'm gonna need help to get it looking like the mockups.

**Comment by ZebulanStanphill** [comment]
Also, I rebased and tested the branch myself and there's no missing border for me. I've tested on the Twenty Twenty theme, though I don't think the theme should make any difference. Could you try again now that the branch has been rebased?

**Comment by mapk** [comment]
I'm still seeing no border in Chrome and Firefox. 🤔 

Screen Shot 2020-07-10 at 10 08 08 AM

**Comment by shaunandrews** [comment]
I dig this, though I think having the space between the parent button and the block toolbar is needed — otherwise its not totally obvious what block is currently selected; Its strange to have two block icons in (visually) single toolbar.

I think you could try adding some margin to the toolbar, and then `absolute` position the parent button, similar to how it works currently.

**Comment by ZebulanStanphill** [comment]
Good idea, @shaunandrews! I did just that, and here's what it looks like now:

image

The CSS is a bit more complicated now, but it's still better than what's in `master`, so I think this will be good to merge if everyone likes how this latest iteration looks/functions.

File: packages/block-editor/src/components/block-toolbar/utils.js
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/block-toolbar/utils.js:31]: This is now only being used for the block outline shown when the block icon is hovered. I could follow up with another PR to remove the debouncing entirely, if desired.

**Review by draganescu — DISMISSED** [review]
I have tested this and the code LGTM 🚢

**Comment by mtias** [comment]
This proposal gives me a lot of pause as it turns something that was _convenient but not primary_ into the most prominent aspect of the block toolbar, visible at all times. Considering nested blocks are going to become more and more the norm, I think we need to be more careful here:

- It's harder to see at a glance what block you are working with.
- The drag / move / transform is pushed into a secondary role.
- Significantly increases the horizontal space the toolbar takes with an action that is not primary to the current block. (More problematic the narrower the context becomes.)
- Select mode doesn't match anymore (the start of the current block type name differs between the two).
- Other cases with child blocks (like horizontal ones: social icons, buttons, etc) become more confusing in terms of alignment:

image

And we get out of bounds much more easily:

image

Given this has the potential of significantly changing the experience by introducing a permanently visible item, I think we should wait a bit more and continue to explore possible solutions.

**Comment by ZebulanStanphill** [comment]
The out-of-bounds thing isn't supposed to happen. I think that issue happens on `master` too? In my testing, resizing the viewport causes the toolbar to get shifted over to not go out of bounds.

The other points are definitely valid, though. Feel free to post suggestions in #23766.

>Other cases with child blocks (like horizontal ones: social icons, buttons, etc) become more confusing in terms of alignment

The iteration without the gap between the parent selector and the rest of the toolbar doesn't have this problem, but it still has all the other issues mentioned, unfortunately.

**Comment by ZebulanStanphill** [comment]
I've been thinking about this for a while. I think a proper solution needs to appear as part of the toolbar (because it is), rather than appear detached from it. And I definitely still think there's value in making the parent-selector more prominent. There's an increasing number of cases where users will want to quickly access the controls of the current parent:

- Get to the Group or Column from the blocks nested in them.
- Get to the Columns from a Column.
- Eventually, getting to the Quote block from the Paragraph blocks nested inside it.
- If we go with an Accordions block implementation that groups multiple individual accordion items so that opening one closes the other, we need to make it easy to access the accordion group block, which is where most of the style controls will probably belong.

Always showing the parent selector when available also makes it more obvious that you're inside a nested block in the first place. This is important info to know when dealing with complex layouts or newly-installed blocks.

The problem with my initial design in this PR is that the difference/relation between the two icons is not immediately clear. What's the difference between the two? I think part of the issue stems from the current block icon button being rather "loaded": it visually conveys state (which block am I editing?) while also serving as the button to open the transforms/styles dropdown. This is questionable from an a11y (and just general usability) perspective. Throwing in another button that uses a block icon is difficult because now you have two buttons with very different actions but a similar appearance.

A full solution would probably involve splitting the current block icon button into two separate things: a simple block icon (with an `aria-label` containing the block name) and a dedicated style/transform button. That would be the best solution, a11y-wise. But that would lengthen the toolbar, running into another longstanding issue with the block toolbar design: namely that there simply isn't enough room to show all the controls properly.

**I've noticed that we keep running into this same issue over and over and over again.** The drag handle shouldn't be invisible like it is now, but making it visible would lengthen the toolbar. For ideal a11y, the movers shouldn't be stacked, but changing that would lengthen the toolbar. It would be good to have a "visible text labels" option, but that would be really difficult to handle because of how much it would lengthen the toolbar.

But while solving these issues is essential in the long run, if you ask me... it's also way outside the scope of this issue. So either we have to solve the major problems with the block toolbar design first and leave this on hold, or this PR needs to come up with something else that is at least better than what we have in `master` right now.

With that in mind, here's an idea I've been sitting on for a while:
image

The chevron-style shape is intended to make it look a bit like a sort of breadcrumb, which helps to clarify the relation between the two icons. I'm not entirely sure how to actually pull this off; I guess it would involve an absolutely-positioned `::after` pseudo-element with a border on two adjacent sides that's been rotated 45 degrees.

What do you think? Is this design "good enough"?

**Comment by ZebulanStanphill** [comment]
I've been thinking about this some more, and I don't think the mockup I posted in the previous comment is "good enough". The more I looked into implementing it, the more complex I found it to be. It would be really, really hard to make it look right. Additionally, the chevron shape never explicitly tells the user anything; it's all based around hoping the user will understand what it means. That's not good UI design. It also introduces a unique design element for no real reason other than trying to work around existing issues.

I've come to the resolution that we need to actually solve the block icon/switcher issue first; we need to split the block icon from the block switcher and give the switcher a more suitable icon. Only then will it be feasible to move the parent selector button into the standard part of the block toolbar while reducing potential confusion.

I'm marking this PR as blocked for the time being.

**Comment by ZebulanStanphill** [comment]
I've removed the gap between the parent-selector-button and the rest of the toolbar. We ought to be avoiding special-case styling like that.

Here's what it looks like after rebasing the PR to include all the latest UI changes in `master`:
image

The inconsistent usage of block icons to mean two different things is still a problem. I'm going to post some ideas about how to move forward in #23766, but for now I wanted to go ahead and push an update to this PR to prevent it from getting too stale.

**Comment by ZebulanStanphill** [comment]
Following my [latest idea](https://github.com/WordPress/gutenberg/issues/23766#issuecomment-697827491) in #23766, I've once again changed how the "Select parent" option is presented. This time, I've moved it out of the main part of the block toolbar and into the "More options" menu.

image

I believe this approach solves every problem with prior designs except, perhaps, the issue of discoverability. Please let me know what you think about this design in #23766.

**Comment by ZebulanStanphill** [comment]
Per @mtias' [feedback](https://github.com/WordPress/gutenberg/issues/23766#issuecomment-700126138), I've added the parent block's icon to the menu item:

image

**Comment by mtias** [comment]
I'm looping back on this one. Thanks for all the iterations and tweaks so far, much appreciated!

I'm coming around to the fact perhaps we need slightly different treatments for some of these contexts. For example, the initial design proposed here matches well for this use case: #28522. Likely also for Gallery. The case of social icons and buttons is a bit tricky since the container is really there to facilitate creation of multiple items, so for that one I feel we need to work on the `absorb toolbar` part and make it seem like there's no parent block at all.

Unsure if we should be applying the same treatment to parents that are indifferent to its children (Cover, Group, Columns, Template Parts). So my current thinking is to split the work in three:

- Use the initial design for some blocks. (Figure out how to do the opt in.)
- Add the ellipsis menu action regardless as it seems useful in general.
- Work on the absorb toolbar design.

**Comment by mtias** [comment]
In terms of inner blocks that would leave us with three configurations:

- Default (show parent icon to the left).
- Absorb Toolbar (combines and simplifies the two toolbars).
- Transparent (limited or no UI for the parent). #7694 

And also include the “select parent” in the ellipsis menu from this PR for mobile and redundancy.

**Comment by ZebulanStanphill** [comment]
@mtias With #28598 merged, do you think it still makes sense to add an overflow menu option to select the parent block?

**Comment by mtias** [comment]
@ZebulanStanphill sorry missed this comment. I think in mobile the situation is still not great and the ellipsis button could be useful, but it's definitely less pressing than before with all the latest tools (parent selector, list view, etc). Another place from where I think it should be possible to select a parent container is the sidebar inspector, particularly on blocks that are semantically tied (buttons:button, navigation:navigation-link, social-icons:social-icon).

**Comment by ZebulanStanphill** [comment]
I forgot this PR was still open. Closing, since #40105 superseded it.

**Pull request closed without merging**
